### ISSUE TITLE
`yarn publish` runs prepare script *after* publishing to npm

### ISSUE THREAD

**Issue by AgentME** [yarnpkg/yarn#4634]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

I have a project (https://github.com/StreakYC/react-smooth-collapse/) which has a prepare script which erases the js directory, then repopulates it with the output of running babel on the src directory. I made some changes, tested (with browserify and jest, not by using the prepare script) and committed them, and then wanted to publish an update to npm. At this time, my js directory still has the old compiled code.

I ran yarn publish, expecting it to run the prepare script and then publish the results:

```
[chris ~/Coding/react-smooth-collapse]$ yarn publish
yarn publish v1.1.0
[1/4] Bumping version...
info Current version: 1.3.0
question New version: 1.3.1
info New version: 1.3.1
[2/4] Logging in...
[3/4] Publishing...
$ rimraf js && babel -s inline -d js/ src/ && flow-copy-source -v src js
src/getTransitionTimeMs.js -> js/getTransitionTimeMs.js
src/index.js -> js/index.js
src/getTransitionTimeMs.js -> js/getTransitionTimeMs.js.flow
src/index.js -> js/index.js.flow
success Published.
[4/4] Revoking token...
info Not revoking login token, specified via config file.
✨  Done in 5.89s.
```

It looks like it successfully upped the version to 1.3.1, ran the prepare script, and published to npm. If I check my js directory, it does have the newly-compiled code. However, if I go to a new directory and `npm install react-smooth-collapse@1.3.1`, then I see that the old js directory is published on npm. (Easy way to tell: the new js/index.js does not contain the string "_mainElSetter" anywhere, which was intended to be in 1.3.1.)

**Please mention your node.js, yarn and operating system version.**

```
[chris ~/Coding/streak/MailFoo]$ node -v
v8.6.0
[chris ~/Coding/streak/MailFoo]$ yarn --version
1.1.0
[chris ~/Coding/streak/MailFoo]$ uname -a
Darwin Chriss-MacBook-Air.local 15.6.0 Darwin Kernel Version 15.6.0: Sun Jun  4 21:43:07 PDT 2017; root:xnu-3248.70.3~1/RELEASE_X86_64 x86_64
```

**Comment by AgentME** [comment]
Apparently I also had the same issue with the last time I published the package react-float-anchor. I just found that the latest version published on npm has the compiled output from the previous release. These are the only two times so far that I've used `yarn publish`.

I've tested to see whether it always happens for me, and so far it seems that the answer is yes. Here's the log of me demonstrating the issue. I create a package which in the prepare script writes the current timestamp to stdout and "prepare.txt". I run `yarn` once so the prepare script runs once. Then I run `yarn publish`. You can see prepare.txt is updated.

Then I go to a new directory, I npm-install the new package, look at the prepare.txt, and I see the timestamp from when I first ran `yarn`, *not* the timestamp generated from when I ran `yarn publish`.

```
[chris ~]$ mkdir /tmp/yarn-publish-test
[chris ~]$ cd /tmp/yarn-publish-test/
[chris /tmp/yarn-publish-test]$ yarn init -y
yarn init v1.1.0
warning The yes flag has been set. This will automatically answer yes to all questions which may have security implications.
success Saved package.json
✨  Done in 0.04s.
[chris /tmp/yarn-publish-test]$ nano package.json 
[chris /tmp/yarn-publish-test]$ cat package.json 
{
  "name": "yarn-publish-test",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "prepare": "( echo prepare && date ) | tee prepared.txt"
  }
}
[chris /tmp/yarn-publish-test]$ yarn
yarn install v1.1.0
info No lockfile found.
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
info Lockfile not saved, no dependencies.
$ ( echo prepare && date ) | tee prepared.txt
prepare
Thu Oct  5 15:26:13 PDT 2017
✨  Done in 0.12s.
[chris /tmp/yarn-publish-test]$ cat prepared.txt 
prepare
Thu Oct  5 15:26:13 PDT 2017
[chris /tmp/yarn-publish-test]$ yarn publish
yarn publish v1.1.0
[1/4] Bumping version...
info Current version: 1.0.0
question New version: 
[2/4] Logging in...
[3/4] Publishing...
$ ( echo prepare && date ) | tee prepared.txt
prepare
Thu Oct  5 15:26:24 PDT 2017
success Published.
[4/4] Revoking token...
info Not revoking login token, specified via config file.
✨  Done in 2.88s.
[chris /tmp/yarn-publish-test]$ cat prepared.txt 
prepare
Thu Oct  5 15:26:24 PDT 2017
[chris /tmp/yarn-publish-test]$ mkdir /tmp/yarn-check
[chris /tmp/yarn-publish-test]$ cd !$
cd /tmp/yarn-check
[chris /tmp/yarn-check]$ npm i yarn-publish-test
npm WARN saveError ENOENT: no such file or directory, open '/private/tmp/yarn-check/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/private/tmp/yarn-check/package.json'
npm WARN yarn-check No description
npm WARN yarn-check No repository field.
npm WARN yarn-check No README data
npm WARN yarn-check No license field.

+ yarn-publish-test@1.0.0
added 1 package in 1.253s
[chris /tmp/yarn-check]$ cat node_modules/yarn-publish-test/prepared.txt 
prepare
Thu Oct  5 15:26:13 PDT 2017
[chris /tmp/yarn-check]$ npm unpublish yarn-publish-test
npm ERR! Refusing to delete entire project.
npm ERR! Run with --force to do this.
npm ERR! npm unpublish [<@scope>/]<pkg>[@<version>]
[chris /tmp/yarn-check]$ npm unpublish --force yarn-publish-test
npm WARN using --force I sure hope you know what you are doing.
npm ERR! registry error parsing json
npm ERR! code E403
npm ERR! You do not have permission to publish "yarn-publish-test". Are you logged in as the correct user? : 1-a69abfb7b2bfb0a3818e2641d23fc530

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/chris/.npm/_logs/2017-10-05T22_28_03_597Z-debug.log
[chris /tmp/yarn-check]$ node -v
v8.6.0
[chris /tmp/yarn-check]$ yarn --version
1.1.0
```

**Comment by AgentME** [comment]
I just tested prepublishOnly and found that it's also run after the files have already been published to npm. (Different but related relevant issue: https://github.com/yarnpkg/yarn/issues/3992)

**Comment by BYK** [comment]
I think this would be fixed by #4553. Am I right?

**Comment by AgentME** [comment]
Just tried it again, looks like it's fixed in yarn v1.2.0!
